server/api/src/main/java/org/hawkular/apm/server/api/utils/ProducerInfoUtil.java

```diff
@@ -156,91 +145,68 @@ protected void initialiseProducerInfo(List<ProducerInfo> producerInfoList, Strin
                 initialiseProducerInfo(producerInfoList, tenantId, trace, origin,
-                        ((ContainerNode) node).getNodes().get(j));
+                        ((ContainerNode) node).getNodes().get(j), producerInfoCache);
             }
         }
     }
 
     /**
-     * This method initialises producer information associated with the supplied
-     * traces.
+     * This method attempts to derive the Producer Information for the supplied server
+     * span. If the information is not available, then a null will be returned, which
+     * can be used to trigger a retry attempt if appropriate.
      *
-     * @param tenantId
-     * @param items
-     * @throws RetryAttemptException Failed to initialise producer information
+     * @param tenantId The tenant id
+     * @param serverSpan The server span
+     * @param spanCache The cache
+     * @return The producer information, or null if not found
      */
-    public void initialiseFromSpans(String tenantId, List<Span> items) throws RetryAttemptException {
-        List<ProducerInfo> producerInfoList = new ArrayList<ProducerInfo>();
-
-        // Identify source URI details
-        // NOTE: This mechanism assumes the list of spans contains all of the spans
-        // associated with a service invocation.
-        Map<String,String> sourceURIs = new HashMap<String,String>();
-        Map<String,String> sourceOps = new HashMap<String,String>();
-        for (int i = 0; i < items.size(); i++) {
-            Span span = items.get(i);
-            if (span.serverSpan()) {
-                URL url = span.url();
-                if (url != null) {
-                    sourceURIs.put(span.getId(), url.getPath());
-
-                    String op = span.operation();
-                    if (op != null) {
-                        sourceOps.put(span.getId(), op);
-                    }
-                }
+    public static ProducerInfo getProducerInfo(String tenantId, Span serverSpan, SpanCache spanCache) {
+        String clientSpanId = SpanUniqueIdGenerator.getClientId(serverSpan.getId());
+        if (spanCache != null && clientSpanId != null) {
+            Span clientSpan = spanCache.get(tenantId, clientSpanId);
+
+            // Work up span hierarchy until find a server span, or top level span
+            Span current = clientSpan;
+            while (current != null &&
+                    !current.serverSpan() && !current.topLevelSpan()) {
+                current = spanCache.get(tenantId, current.getParentId());
             }
-        }
-
-        // This method initialises the deriver with a list of trace fragments
-        // that will need to be referenced when correlating a consumer with a producer
-        for (int i = 0; i < items.size(); i++) {
-            // Need to check for Client spans
-            Span span = items.get(i);
 
-            if (span.clientSpan()) {
+            if (current != null) {
```

Comment:
Not sure what you mean? It represents the current node while walking up the hierarchy until either the top level node or a server span is reached. So at the end it will represent a root or server span, but not until it has walked up the hierarchy.